STM/vertx/README.md

```diff
@@ -0,0 +1,181 @@
+--------
+
+Another example of STM and Vert.x integration. The quickstart is divided into 4 parts each introducing
+a different facet of STM and Vert.x integration.
+
+Demo 1 shows a single JVM running multiple vertx instances sharing the same volatile STM object.
+
+Demo 2 uses the same STM object but makes it persistent and shows how to share it across address
+spaces.
+
+These two examples illustrate the different ways in which an application can be scaled whilst
+maintaining data consistency:
+
+ - verticle scaling by using better hardware so that more threads can do be used to service the
+   workload;
+ - horizontal scaling by using more servers so that the workload can be distributed to multiple JVMs
+
+Demos 3 and 4 show services running on different endpoints communicating via HTTP and shared memory
+respectively.
+
+First build a fat jar that contains all the classes needed to run both demos in a single jar:
+
+  mvn clean package
+
+Use Case 1: Scaling by adding more threads
+==========================================
+
+  Charactertics:
+
+  - RECOVERABLE and EXCLUSIVE
+  - creates multiple vertx instances each instance using a handle to the same STM object
+  
+  What it's good for:
+
+  - vertical scaling where adding better h/w is an option in order to support more threads in one JVM
+
+  Features:
+
+   This example shows a Theatre booking service:
+   - an STM object is used to maintain all Theatre bookings
+     * the STM object is volatile
+   - multiple vertx instances each listening on the same HTTP endpoint
+   - each vertx instance shares the same STM object
+   - all vertx instances run in the same address space
+     * concurrency is managed by the STM runtime
+   - shows vertical scaling
+
+Usage:
+------
+
+  java -cp target/stm-vertx-demo-5.6.0.Final-SNAPSHOT-fat.jar demo.demo1.VolatileTheatreVerticle 
+
+  This will start You can change the port (8080) on which the verticle event loops will listen and the
+     number of instances (10) of the verticles. Note that the version of the jar (5.6.0.Final-SNAPSHOT
+     in this example) will change depending upon which narayana release you have.
+
+Create 2 bookings:
+
+  curl -X POST http://localhost:8080/api/theatre/A
+  curl -X POST http://localhost:8080/api/theatre/A
+
+observe how each request is serviced on a different verticle instance
+
+Similarly performing GETs will show the same booking counts regardless of which verticle services it:
+
+  curl -X GET http://localhost:8080/api/theatre 
+  curl -X GET http://localhost:8080/api/theatre 
+
+Use Case 2: Scaling by distributing the workload across JVMs
+==========================================
+
+Similar to use case 1 but uses persistent STM objects spread across different JVMs.
+
+  Charactertics:
+
+  - PERSISTENT and SHARED
+  - theatre service verticles running in different JVMs sharing the same STM object
+  - each JVM hosting multiple vertx instances each instance using a handle to the same STM object
```

Comment:
- Although not demonstrated here, there is an assumption that each process has access to the same object store